### PR TITLE
Feature/gazebo9 noisydepth

### PR DESCRIPTION
**Pull request by michaelpantic**
- Simple plugin that adds a depth sensor with noise
- Includes two noise-depth models
- Kinect model is a simplification of https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=6375037
- D435 model is the "theoretical achievable RMS" model from the intel whitepaper "Best-Known-Methods for Tuning Intel RealSense™ D400 Depth Cameras for Best Performance"

- Will probably add better models in the future.

If needed, I can also do a backport to Pre-Gazebo9. 
Anyway, we maybe should think about the Gazebo9 issues, if we do a hard-fork or not - it becomes an issue ;-)

### CODE REVIEW

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_noisydepth_plugin.h [hunk 1]
```diff
@@ -0,0 +1,160 @@
+/*
```
Comment (ffurrer) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_noisydepth_plugin.h]: I think I would slightly prefer gazebo_noisy_depth_plugin, but you can also keep this ;)

File: rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+#include <algorithm>
+#include <assert.h>
+#include <boost/thread/thread.hpp>
+#include <boost/bind.hpp>
+
+#include <rotors_gazebo_plugins/gazebo_noisydepth_plugin.h>
+
+#include <gazebo/sensors/Sensor.hh>
+#include <sdf/sdf.hh>
+#include <gazebo/sensors/SensorTypes.hh>
```
Comment (ffurrer) [inline rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp]: alphabetize

File: rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+  this->format_ = this->format;
+  this->camera_ = this->depthCamera;
+
+  // using a different default
+  if (!_sdf->HasElement("imageTopicName"))
+    this->image_topic_name_ = "ir/image_raw";
+  if (!_sdf->HasElement("cameraInfoTopicName"))
+    this->camera_info_topic_name_ = "ir/camera_info";
+
+
```
Comment (ffurrer) [inline rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp]: rm vspace (just run clang-format also on this file :))

File: rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+  this->parentSensor_ = this->parentSensor;
+  this->width_ = this->width;
+  this->height_ = this->height;
+  this->depth_ = this->depth;
+  this->format_ = this->format;
+  this->camera_ = this->depthCamera;
+
+  // using a different default
+  if (!_sdf->HasElement("imageTopicName"))
+    this->image_topic_name_ = "ir/image_raw";
```
Comment (ffurrer) [inline rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp]: can you make your if clauses:

```C++
if (cond) {
  ...
}
```

File: rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+////////////////////////////////////////////////////////////////////////////////
+// Increment count
+void GazeboNoisyDepth::DepthInfoConnect() {
+  this->depth_info_connect_count_++;
+}
+////////////////////////////////////////////////////////////////////////////////
+// Decrement count
+void GazeboNoisyDepth::DepthInfoDisconnect() {
+  this->depth_info_connect_count_--;
```
Comment (ffurrer) [inline rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp]: pre(in/de)crement here and above

File: rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+      this->parentSensor->SetActive(false);
+    } else {
+      if (this->depth_image_connect_count_ > 0)
+        this->FillDepthImage(_image);
+    }
+  } else {
+    if (this->depth_image_connect_count_ <= 0)
+      // do this first so there's chance for sensor to run 1 frame after activate
+      this->parentSensor->SetActive(true);
+  }
```
Comment (ffurrer) [inline rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp]: This seems a bit complicated, I guess the important part is filling the depth image when `depth_image_connect_count_ > 0 && this->parentSensor->IsActive()` and then you want a functionality do activate or deactivate the parent sensor. (But this part I didn't fully get what you wanted to do)

File: rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+      this->parentSensor->SetActive(false);
+    } else {
+      if ((*this->image_connect_count_) > 0)
+        this->PutCameraData(_image);
+    }
+  } else {
+    if ((*this->image_connect_count_) > 0)
+      // do this first so there's chance for sensor to run 1 frame after activate
+      this->parentSensor->SetActive(true);
+  }
```
Comment (ffurrer) [inline rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp]: Same as above

File: rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+
+////////////////////////////////////////////////////////////////////////////////
+// Update the controller
+void GazeboNoisyDepth::OnNewImageFrame(const unsigned char *_image,
+                                       unsigned int _width, unsigned int _height, unsigned int _depth,
+                                       const std::string &_format) {
+  if (!this->initialized_ || this->height_ <= 0 || this->width_ <= 0)
+    return;
+
+  //ROS_ERROR_NAMED("openni_kinect", "camera_ new frame %s %s",this->parentSensor_->Name().c_str(),this->frame_name_.c_str());
```
Comment (ffurrer) [inline rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp]: rm?

File: rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+    }
+  } else {
+    if ((*this->image_connect_count_) > 0)
+      // do this first so there's chance for sensor to run 1 frame after activate
+      this->parentSensor->SetActive(true);
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Put depth image data to the interface
```
Comment (ffurrer) [inline rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp]: This belongs rather to the header file :) (and is probably not adding so much information)

File: rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+void GazeboNoisyDepth::FillDepthImage(const float *_src) {
+  this->lock_.lock();
+  // copy data into image
+  this->depth_image_msg_.header.frame_id = this->frame_name_;
+  this->depth_image_msg_.header.stamp.sec = this->depth_sensor_update_time_.sec;
+  this->depth_image_msg_.header.stamp.nsec = this->depth_sensor_update_time_.nsec;
+
+  ///copy from depth to depth image message
```
Comment (ffurrer) [inline rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp]: fix comments also in this file (also check `//` vs `///`)

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_noisydepth_plugin.h [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+  int depth_image_connect_count_;
+ private:
+  void DepthImageConnect();
+ private:
+  void DepthImageDisconnect();
+
+ private:
+  bool FillDepthImageHelper(sensor_msgs::Image &image_msg,
+                            uint32_t height, uint32_t width,
+                            uint32_t step, void *data_arg);
```
[the remark this answers is not in the document]
Comment (ffurrer) [reply rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_noisydepth_plugin.h]: ahh no, just realized, this actually fills in the image_msg so move image_msg to end and make it a pointer. (https://google.github.io/styleguide/cppguide.html#Reference_Arguments)

File: rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+  this->depth_image_pub_.publish(this->depth_image_msg_);
+
+  this->lock_.unlock();
+}
+
+// Fill depth information
+bool GazeboNoisyDepth::FillDepthImageHelper(
+    sensor_msgs::Image &image_msg,
+    uint32_t rows_arg, uint32_t cols_arg,
+    uint32_t step_arg, void *data_arg) {
```
Comment (ffurrer) [inline rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp]: try to pass data_arg in by const reference and a proper type and move image_msgs to the end and switch to pointer type

File: rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+  ROS_DEBUG_NAMED("NoisyDepth", "publishing default camera info, then openni kinect camera info");
+  GazeboRosCameraUtils::PublishCameraInfo();
+
+  if (this->depth_info_connect_count_ > 0) {
+    this->sensor_update_time_ = this->parentSensor_->LastMeasurementTime();
+#if GAZEBO_MAJOR_VERSION >= 8
+    common::Time cur_time = this->world_->SimTime();
+#else
+    common::Time cur_time = this->world_->GetSimTime();
+#endif
```
Comment (ffurrer) [inline rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp]: Do your new version agnostic magic ;) but make a separate PR for this :)

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/depth_noise_model.hpp [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+class D435DepthNoiseModel : public DepthNoiseModel {
+
+ public:
+
+  D435DepthNoiseModel() :
+      h_fov(M_PI_2),     // Default 90deg for D435
+      baseline(0.05),    // Default 50 mm for D435
+      subpixel_err(0.1), // Default subpixel calibration error
+      max_stdev(3.0),
```
[the remark this answers is not in the document]
Comment (michaelpantic) [reply rotors_gazebo_plugins/include/rotors_gazebo_plugins/depth_noise_model.hpp]: Kept float, as data in this case is inherently float.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_noisydepth_plugin.h [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+  /// \brief Advertise point cloud and depth image
+ public:
+  virtual void Advertise();
+
+  /// \brief Update the controller
+ protected:
+  virtual void OnNewDepthFrame(const float *_image,
+                               unsigned int _width, unsigned int _height,
+                               unsigned int _depth, const std::string &_format);
```
[the remark this answers is not in the document]
Comment (michaelpantic) [reply rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_noisydepth_plugin.h]: Interface fixed by DepthCameraPlugin (this is inherited).

File: rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp (hunk 6 above)
Comment (michaelpantic) [reply rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp]: as the logic is somewhat duplicate between OnNewImageFrame and OnNewDepthFrame it's confusing. In one of them it was actually implemented correctly (IMHO). 
The second check should check if there are image subscribers on the parent (inherited property image_connect_count) and if so, activate the parent sensor, before outputing sensor data itself.

Fixed it and commented it a bit. Seems to be ok.

**Comment by michaelpantic** [comment]
Actually wanted to avoid reformating/changing/fixing the code I based this on, as it is in use since years and clearly works. Anyway, here we go.

Btw most of the weirdness in the original code seemed to be in line with the rest of the gazebo source code (e.g. specifying access modifiers for each member individually, etc).

**Comment by michaelpantic** [comment]
Keeping branch for now, as I might do backport.
